[st7123] add ST7123 touch controller component (M5Stack Tab5)

miniskipper opened the pull request:
# What does this implement/fix?

>Starting from October 14, 2025, the Tab5’s original independent display driver ILI9881C and touch driver GT911 will be replaced by the integrated display‑touch driver ST7123. Some early firmware builds may not run properly. The latest versions of M5Unified and M5GFX have already been adapted for compatibility with this new screen driver, and older programs can be recompiled using the latest M5Unified and M5GFX to achieve proper compatibility.
By checking the sticker on the back of the Tab5 product, you can confirm the driver model of the device.

Add support for the touch controller found on the M5Stack Tab5 (newer version)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Code quality improvements to existing code or addition of tests
- [ ] Other

**Related issue or feature (if applicable):**

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):**
- esphome/esphome-docs#5675

## Test Environment

- [ ] ESP32
- [x] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040
- [ ] BK72xx
- [ ] RTL87xx
- [ ] nRF52840

## Example entry for `config.yaml`:

```yaml
touchscreen:
  - platform: st7123
    i2c_id: bsp_bus
    interrupt_pin: GPIO23
    update_interval: never
    reset_pin:
      pi4ioe5v6408: pi4ioe1
      number: 5
    calibration:
      x_min: 0
      x_max: 720
      y_min: 0
      y_max: 1280
    id: tab5_touch
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

miniskipper committed:
[st7123] add ST7123 Touch Controller (M5Stack Tab5)

miniskipper committed:
[st7123] read keys register

miniskipper committed:
[st7123] doc

miniskipper committed:
Merge branch 'esphome:dev' into st7123-touch

miniskipper committed:
[st7123] read max/min coordinates from controller

miniskipper committed:
Merge branch 'esphome:dev' into st7123-touch

swoboda1337 reviewed (changes requested):
CI is failing and there are no tests.

miniskipper committed:
Merge branch 'dev' into st7123-touch

miniskipper committed:
[st7123] fomatting, deprecated mark_failed

miniskipper committed:
[st7123] test

miniskipper committed:
Merge branch 'dev' into st7123-touch

miniskipper committed:
Merge branch 'dev' into st7123-touch

miniskipper committed:
Merge branch 'dev' into st7123-touch

miniskipper committed:
Merge branch 'dev' into st7123-touch

AlexMohanMD commented:
@swoboda1337 - can you review this please? this is a blocker for a current project of mine, looks like they requested review again.

clydebarrow commented on esphome/components/st7123/touchscreen/st7123_touchscreen.h:
```diff
@@ -0,0 +1,56 @@
+  /// is called to complete the initialization.
+  void setup() override;
+  void dump_config() override;
+  bool can_proceed() override { return this->setup_done_; }
```
This will block other components' setup() calls and should be removed.

clydebarrow reviewed (changes requested):
Remove the esp8266 and rp2040 test files - there is no value in these and they just increase CI build times.

clydebarrow committed:
Merge branch 'dev' into st7123-touch

clydebarrow committed:
Merge branch 'dev' of https://github.com/esphome/esphome into st7123-touch

clydebarrow committed:
Remove blocking `can_proceed()` call;
Remove redundant tests
Remove binary sensor features - not used by tab5 so currently
untestable.

clydebarrow committed:
Fixes from review

clydebarrow commented on tests/components/st7123/test.esp32-idf.yaml:
```diff
@@ -0,0 +1,9 @@
+substitutions:
+  display_reset_pin: "10"
+  interrupt_pin: "20"
+  reset_pin: "21"
```
Not required, it would add no value.
